Tessellator refactor + floors

ViperMiniQ opened the pull request:
This PR refactors the QgsTessellator to make it more easier to work with and for potential addition of new features to it.

Existing constructors are deprecated and replaced with a default one and setters are introduced to set all the options. 
Instead of passing the full parameter list, we now have setters for setting: 
 - extrusion faces (where we choose what we want to add to the scene, floors, roofs and walls - any combination)
 - inverted normals
 - textures
 - rotation
 - origin

Some of the bigger methods are split into multiple smaller. 

Enum `ExtrusionFace` replaces the facade int, optional creation of floors is added.

New functionality is the optional creation of floors.
Edit: now the options for Rendered facade are: `Walls`, `Walls and Roof`, `Walls, Roof and Floor`

<img width="956" height="631" alt="Screenshot_20260113_124114" src="https://github.com/user-attachments/assets/fdb70856-93d7-42de-9ce1-623abdefe446" />


Funded by: https://www.lutraconsulting.co.uk/crowdfunding/qgis-3d-for-open-source-digital-twins

ViperMiniQ committed:
add default constructor, deprecate old constructors, introduce setters for tessellator options

ViperMiniQ committed:
create private method for adding triangle vertices (remove duplicate code)

ViperMiniQ committed:
move base transform calculation to private method

ViperMiniQ committed:
move addTextureCoords to private method

ViperMiniQ committed:
move triangle creation (poly2tri) to private method

ViperMiniQ committed:
rework addPolygon method to use newly created methods

ViperMiniQ committed:
fix division by zero

ViperMiniQ committed:
update tesellator header

ViperMiniQ committed:
tests should use the default constructor

ViperMiniQ committed:
rework 3d symbol to use enums

ViperMiniQ committed:
use default constructor

ViperMiniQ committed:
remove old values from gui

ViperMiniQ committed:
sipify

ViperMiniQ committed:
use default constructor

ViperMiniQ committed:
fix warnings

ViperMiniQ committed:
set base transform calculation to const

ViperMiniQ committed:
move texture coords addition in addtrianglevertices

ViperMiniQ committed:
move polygon transform at beginning, use extrusionorigin when adding vertices

ViperMiniQ committed:
fix test

ViperMiniQ committed:
add noface in switch

ViperMiniQ committed:
switch to default constructor

ViperMiniQ committed:
use actual polygon point coords

ViperMiniQ committed:
fix clang tidy warnings

ViperMiniQ committed:
add extrusionfaces enum

ViperMiniQ committed:
fix clang tidy cast warning

ViperMiniQ committed:
float

wonder-sk commented on src/3d/processing/qgsalgorithmtessellate.cpp:
```diff
@@ -126,7 +126,8 @@ QgsFeatureList QgsTessellateAlgorithm::processFeature( const QgsFeature &feature
         // 3D case, or 2D case with unsupported GEOS version -- use less stable poly2tri backend
         const QgsRectangle bounds = f.geometry().boundingBox();
-        QgsTessellator t( bounds, false );
+        QgsTessellator t = QgsTessellator();
```
```suggestion
        QgsTessellator t;
```

wonder-sk commented on src/core/vector/qgsvectorlayerprofilegenerator.cpp:
```diff
@@ -1467,7 +1467,8 @@ bool QgsVectorLayerProfileGenerator::generateProfileForPolygons()
       {
         const QgsRectangle bounds = clampedPolygon->boundingBox();
-        QgsTessellator t( bounds, false, false, false, false );
+        QgsTessellator t = QgsTessellator();
```
```suggestion
        QgsTessellator t;
```

uclaros commented on tests/src/3d/testqgstessellator.cpp:
```diff
@@ -438,7 +457,10 @@ void TestQgsTessellator::testBackEdges()
 
-  QgsTessellator tN( 0, 0, true, false, true );
+  // QgsTessellator tN( 0, 0, true, false, true );
+  QgsTessellator tN = QgsTessellator();
```
```suggestion
  QgsTessellator tN;
```
Same in other cases of ` = QgsTessellator()`

wonder-sk commented on src/core/qgstessellator.cpp:
```diff
@@ -175,39 +176,108 @@ static void make_quad( float x0, float y0, float z0, float x1, float y1, float z
+  setExtrusionFacesLegacy( facade );
+  setAddBackFaces( addBackFaces );
+  setAddTextureUVs( addTextureCoords );
+  setExtrusionEnabled( noZ );
```
Actually... we didn't have "extrusion enabled" property before, probably we don't really need it at all :smile:  it is simply per-feature property (depending on whether extrusion height is zero or not).

uclaros commented on src/3d/symbols/qgspolygon3dsymbol_p.cpp:
```diff
@@ -98,11 +98,33 @@ bool QgsPolygon3DSymbolHandler::prepare( const Qgs3DRenderContext &context, QSet
-
-  outNormal.tessellator->setOutputZUp( true );
-  outSelected.tessellator->setOutputZUp( true );
+  QgsTessellator *tessellator = new QgsTessellator();
```
Let's use `make_unique` + `std::move` in these too!

uclaros commented on src/3d/symbols/qgspolygon3dsymbol_p.cpp:
```diff
@@ -98,11 +98,33 @@ bool QgsPolygon3DSymbolHandler::prepare( const Qgs3DRenderContext &context, QSet
+  tessellator->setExtrusionEnabled( true );
+  tessellator->setOutputZUp( true );
+  tessellator->setExtrusionFaces( mSymbol->extrusionFaces() );
+  tessellator->setTextureRotation( texturedMaterialSettings ? static_cast<float>( texturedMaterialSettings->textureRotation() ) : 0 );
```
clang-tidy might cry (also for the outNormal tessellator below)
```suggestion
  tessellator->setTextureRotation( texturedMaterialSettings ? static_cast<float>( texturedMaterialSettings->textureRotation() ) : 0.f );
```

uclaros commented on src/core/qgstessellator.cpp:
```diff
@@ -473,337 +543,275 @@ double _minimum_distance_between_coordinates( const QgsPolygon &polygon )
+
+  for ( p2t::Triangle *t : triangles )
+  {
+    trianglePoints.emplace_back( QVector3D( t->GetPoint( 0 )->x / mScale, t->GetPoint( 0 )->y / mScale, z.value( t->GetPoint( 0 ) ) ) );
```
With `emplace_back`, shouldn't these be:
```suggestion
    trianglePoints.emplace_back( t->GetPoint( 0 )->x / mScale, t->GetPoint( 0 )->y / mScale, z.value( t->GetPoint( 0 ) ) );
```
to avoid the copy?

uclaros commented on src/core/qgstessellator.cpp:
```diff
@@ -473,337 +543,275 @@ double _minimum_distance_between_coordinates( const QgsPolygon &polygon )
-  const QVector3D upVector( 0, 0, 1 );
-  const float pNormalUpVectorDotProduct = QVector3D::dotProduct( upVector, pNormal );
-  const float radsBetweenUpNormal = static_cast<float>( qAcos( pNormalUpVectorDotProduct ) );
+  // const float detectionDelta = qDegreesToRadians( 10.0f );
```
```suggestion
```
Is this logic with the `detectionDelta` angle not needed anymore?

ViperMiniQ committed:
address review comments

- z values are offset if origin is set with z value

- mNoZ renamed to inputZValuesIgnored with setInputZValueIgnored/isZValueIgnored setters/getters

- add setters

ViperMiniQ commented on src/core/qgstessellator.cpp:
```diff
@@ -175,39 +176,108 @@ static void make_quad( float x0, float y0, float z0, float x1, float y1, float z
+void QgsTessellator::setAddBackFaces( bool addBackFaces )
+{
+  mAddBackFaces = addBackFaces;
+  updateStride();
```
Don't know how that got there...

ViperMiniQ committed:
double to float conversion

ViperMiniQ commented on src/core/qgstessellator.h:
```diff
@@ -51,9 +56,65 @@ class CORE_EXPORT QgsTessellator
+     * \since QGIS 4.0
+     */
+    void setInvertNormals( bool invertNormals );
 
```
`setExtrusionEnabled` (yes, a bad name), renamed to `setInputZValuesIgnored` with setters and getters.

wonder-sk commented on src/core/qgstessellator.h:
```diff
@@ -51,9 +56,114 @@ class CORE_EXPORT QgsTessellator
+    Qgis::ExtrusionFaces extrusionFaces() const { return mExtrusionFaces; }
+
+    /**
+     * Sets the rotation of generated faces.
```
```suggestion
     * Sets the rotation of texture UV coordinates (in degrees).
```
(also for getter)

wonder-sk commented on src/core/qgstessellator.h:
```diff
@@ -51,9 +56,114 @@ class CORE_EXPORT QgsTessellator
+     * Returns whether back faces are being added to the output data (TRUE) or not (FALSE).
+     * \since QGIS 4.0
+     */
+    bool hasBackFaces() const { return mAddBackFaces; }
```
it would be good to have consistent name with the setter - here we miss the "Add" bit, which would indeed read a bit weird here. How about setBackFacesEnabled() / hasBackFacesEnabled() ?

wonder-sk commented on src/core/qgstessellator.cpp:
```diff
@@ -175,39 +176,107 @@ static void make_quad( float x0, float y0, float z0, float x1, float y1, float z
+    case 3:
+      mExtrusionFaces = Qgis::ExtrusionFace::Walls | Qgis::ExtrusionFace::Roof;
+      break;
+    case 4:
```
this should be 7 to be correct

ViperMiniQ committed:
2nd review

- remove reduntant NaN checks
- rename some setters/getters
- use origin Z for normal calculation

ViperMiniQ committed:
revert accidental rename

ViperMiniQ committed:
fix reserved in the global namespace warning

DelazJ commented:
@ViperMiniQ It looks like this PR modified at least the "r[endered facade](https://docs.qgis.org/testing/en/docs/user_manual/style_library/3d_symbols.html#d-polygonlayers)" option. Can you confirm, and if so, adjust the description for something more digest for writers, please? Thanks
